Updated Defold for Unity users manual on MonoBehaviour vs Lua modules. (#631)
defold/doc: docs/en/manuals/unity.md

Defold is a completely free, truly cross-platform 3D game engine with an Editor for Windows, Linux, and macOS. The full source code is available on [Github](https://github.com/defold/defold/).

> Defold is focused on performance, even on low-end devices. It uses a small, explicit component model where many gameplay interactions are handled through code and message passing.

Reviewer: What do you mean by "explicit"?

Author: That all components are "explicitly" added (`m_Components`) - isn't that good wording? Do you think we should remove it in order not to make more confusion?

Author: Yeah, I'll remove it.

#### Behaviour components

> In Unity, "component" can mean an engine component or a custom `MonoBehaviour` that you attach to a GameObject. In Defold, the word "Component" usually refers to engine-level components such as Sprite, Collision Object, Factory, GUI, Sound, Model, Script, and so on. A Script is one such component, but Defold projects usually do not treat scripts as one-to-one equivalents of Unity `MonoBehaviour` scripts.

Reviewer: I think Component in Unity is almost always understood to be a MonoBehaviour. To the point where that's what the code API uses; `GetComponent<T>`, `GetComponentInChildren<T>` etc;. I would maybe phrase it as;

```
In Unity, "component" usually means a `MonoBehaviour` that is attached to a GameObject. You can create your own by inheriting from `MonoBehaviour` or you can use their built-in components like Light, some physics stuff, something something, so on and so forth.

In Defold, Component refers exclusively to what would be the equivalent built-in components in Unity, or something, but defold does not treat a script as a monobehaviour and does not require any explicit "marking" in order to attach to a gameobject, other than creating listener events/callbacks.
```

## Code Writing

> In Defold scripts are written in Lua, with a typed dialect of Lua - Teal - coming as next first class supported language. Native Extensions can be written in several other languages: C, C++, C#, Objective-C, Java or JS depending on target platform.

Reviewer: ```suggestion
In Defold, scripts are written in Lua, with a typed dialect of Lua - Teal - coming as next first class supported language. Native Extensions can be written in several other languages: C, C++, C#, Objective-C, Java or JS depending on target platform.
```

### C++/C# Native Extensions

> In Defold you can write Native Extensions can be written in several other languages: C, C++, C#, Objective-C, Java or JS depending on target platform. If you are very comfortable with C#, it’s technically possible to structure most of your game logic in a C# extension and just call it from a small Lua bootstrap script, though this requires some advanced API knowledge and is not recommended for beginners.

Reviewer: In Defold, Native Extensions can be written...

Read more about extensions in [Defold Native Extensions manual](/manual/extensions.md).

### From MonoBehaviours to Lua modules

A common pitfall for developers coming from Unity is treating Defold scripts like `MonoBehaviour` and attaching ones to game objects.

While you can definitely write in an object oriented way, and there are even libraries to help you with this ([defold-oop](https://github.com/xiyoo0812/defold-oop) or [lua-class](https://github.com/d954mas/lua-class)), the recommended way is to use scripts as systems or managers.

> A single script in Defold can even control hundreds or thousands of other objects and their components, while them having no scripts of their own, thanks to powerful addressing and messaging in Defold. Creating a separate script for each object is rarely necessary and can lead to a counterproductive complexity.

Reviewer: ```suggestion
A single script in Defold can even control hundreds or thousands of other objects and their components, even without them having any scripts of their own, thanks to a powerful messaging system in Defold. Creating scripts to match each GameObject is rarely necessary and can lead to a counterproductive complexity.
```

end
```

The important difference is not that Defold prevents modular architecture - it is where composition happens and how gameplay code communicates:

| Unity | Defold |
|---|---|
| Attach several `MonoBehaviour` scripts in the Inspector | Attach one `.script` and compose Lua modules in code |
| Use `GetComponent<T>()` or serialized fields to connect behaviours | Store module instances on `self` and use addresses/messages between objects |
| Each component can have its own lifecycle methods | The host script routes `init()`, `update()`, `on_message()`, `final()`, etc. |
| Many architectural styles are possible | Messaging-oriented, explicit code composition is the common practice |

This can feel unusual at first, especially if you are used to configuring behaviour by adding components in the Inspector. In Defold, many things that you might configure visually in Unity can instead be created, connected, enabled, disabled, or updated through code. Defold's messaging system helps decouple logic: the sender posts data to an address, and the receiver decides what to do with it.

> This approach, even though recommended, is not enforced and you can still write your scripts however you like, including attaching multiple scripts per game object or getting closer to an object oriented programming, and there are even libraries to help you with this ([defold-oop](https://github.com/xiyoo0812/defold-oop) or [lua-class](https://github.com/d954mas/lua-class)).

Reviewer: "or getting closer to an object oriented **style of** programming"
